core/storage/couchStorage.go

```diff
@@ -0,0 +1,691 @@
+}
+
+// Stop stops the CouchStorage store
+func (store *CouchStorage) Stop() {
+	store.client.Close(context.TODO())
+}
+
+// PerformMaintenance performs store's maintenance
+func (store *CouchStorage) PerformMaintenance() {
+	//store.checkObjects()
+}
+
+// Cleanup erase the on disk Bolt database only for ESS and test
+func (store *CouchStorage) Cleanup(isTest bool) common.SyncServiceError {
+	return nil
+}
+
+// GetObjectsToActivate returns inactive objects that are ready to be activated
+func (store *CouchStorage) GetObjectsToActivate() ([]common.MetaData, common.SyncServiceError) {
+	return nil, nil
+}
+
+// StoreObject stores an object
+// If the object already exists, return the changes in its destinations list (for CSS) - return the list of deleted destinations
+func (store *CouchStorage) StoreObject(metaData common.MetaData, data []byte, status string) ([]common.StoreDestinationStatus, common.SyncServiceError) {
+	id := getObjectCollectionID(metaData)
+	var dests []common.StoreDestinationStatus
+	var deletedDests []common.StoreDestinationStatus
+
+	if metaData.DestinationPolicy != nil {
+		metaData.DestinationPolicy.Timestamp = time.Now().UTC().UnixNano()
+	}
+
+	if status == common.NotReadyToSend || status == common.ReadyToSend {
+		// The object was receieved from a service, i.e. this node is the origin of the object:
+		// set its instance id and create destinations array
+		newID := store.getInstanceID()
+		metaData.InstanceID = newID
+		if data != nil && !metaData.NoData && !metaData.MetaOnly {
+			metaData.DataID = newID
+		}
+
+		var err error
+		dests, deletedDests, err = createDestinationsFromMeta(store, metaData)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	existingObject := &couchObject{}
+	if err := store.getOne(id, existingObject); err != nil {
+		if !strings.HasPrefix(err.Error(), "Not Found") {
+			return nil, &Error{fmt.Sprintf("Failed to retrieve object's status. Error: %s.", err)}
+		}
+		existingObject = nil
+	}
+
+	if existingObject != nil {
+		if (metaData.DestinationPolicy != nil && existingObject.MetaData.DestinationPolicy == nil) ||
+			(metaData.DestinationPolicy == nil && existingObject.MetaData.DestinationPolicy != nil) {
+			return nil, &common.InvalidRequest{Message: "Can't update the existence of Destination Policy"}
+		}
+
+		if metaData.MetaOnly {
+			metaData.DataID = existingObject.MetaData.DataID
+			metaData.ObjectSize = existingObject.MetaData.ObjectSize
+			metaData.ChunkSize = existingObject.MetaData.ChunkSize
+		}
+		if metaData.DestinationPolicy != nil {
+			dests = existingObject.Destinations
+		}
+
+	}
+
+	newObject := couchObject{ID: id, MetaData: metaData, Status: status, PolicyReceived: false,
+		RemainingConsumers: metaData.ExpectedConsumers,
+		RemainingReceivers: metaData.ExpectedConsumers, Destinations: dests}
+
+	if existingObject != nil {
+		if err := store.upsert(newObject, true); err != nil {
+			return nil, &Error{fmt.Sprintf("Failed to store an object. Error: %s.", err)}
```

Comment:
ah, because of object revisions in couch, you had to discriminate the upsert behavior. You could split upsert into separate add and update functions (and then get rid of upsert). As it is now you have 2 paths through both functions (the end of storeObject and upsert).

Reply:
Okay sure, I will do that. Is it a bad design practice to introduce new variables like exists (bool) in this case? Or is it because the whole point of an 'upsert' function was the calling function not knowing if the object already exists, but in the case of couch we would need to know that to set the boolean value, so it makes more sense to have separate functions?

> As it is now you have 2 paths through both functions (the end of storeObject and upsert)

 I'm sorry I didn't understand what you meant by this, could you please clarify? Thank you